### ISSUE TITLE
Mutation `updateSimpleInventoryBulk` does not update `inventoryReserved`

### ISSUE THREAD

**Issue by zenweasel** [reactioncommerce/reaction#5791]
This method is supposed to both update the `inventoryInStock` and update the `inventoryReserved` but currently does not update `inventoryReserved`

## To Reproduce

1. Use a product that had orders for it but that order is now shipped
1. Issue a call to this mutation that updates the inventory
1. Observe that the `reservedInventory` quantity does not update

**Comment by chrispotter** [comment]
thanks for the issue @zenweasel. 

Is this something that just stopped working recently?

From the core `updateSimpleInventory` and `updateSimpleInventoryBulk` I see a comment 
```
  // If we inserted, set the "reserved" quantity to what it should be. We could have
  // put this in the $setOnInsert but then we'd have to do the Orders lookup for
  // calculating reserved every time, even when only an update happens. It's better
  // to wait until here when we know whether we inserted.
```
There are a few assumptions that I want to clarify.

Right now only orders that emit the `afterOrderApprovePayment` event will have the `inventoryReserved` amount recalculated.  

We do have a note here that says
```
// TODO: When the approval step is removed, change this to afterFulfillmentGroupPacked
  appEvents.on("afterOrderApprovePayment", async ({ order }) => {
```
Are you sure this event is being emitted?

In `recalculateReservedInventory` we are updating this `reservedQuantity` based off when an order is marked as `approved`, which comes from this code
```
export default function orderIsApproved(order) {
  return !Array.isArray(order.payments) ||
    order.payments.length === 0 ||
    !order.payments.find((payment) => payment.status === "created");
}
```
Perhaps the inventoryImport is recalculating inventory but just returning the same value...

So long story short, the reproduction step `Use a product that had orders for it but that order is now shipped` has no bearing on the recalculation of the `reservedInventory` if your payment methods are still set to `created`.

If your reserved inventory is based on if a product is shipped, this logic on `approvedOrders` might need to be modified so account for shipments rather than payments.

This is my understanding so far but maybe @aldeed can chime in as well.

**Comment by zenweasel** [comment]
This appears to have never worked but it just became obvious how that we have a significant number of orders

**Comment by zenweasel** [comment]
@chrispotter Thanks for your comments and looking at this issue, it was very helpful.

Ok, here's the part that gets confusing:

`getReservedQuantity` called from `updateSimpleInventory` is here: 

https://github.com/sportsdirect/reaction/blob/5a2ce1c47fc22fd8474fda14643888ae169cab88/imports/plugins/included/simple-inventory/server/no-meteor/utils/getReservedQuantity.js#L17

And looks like this:
```
  const orders = await context.collections.Orders.find({
    "workflow.status": { $in: ["new", "coreOrderWorkflow/processing"] },
    "shipping.items.variantId": productVariantId
  }).toArray();
```

So I am going off the logic in this mutation which is title `getReservedQuanttiy` which does not look at payments at all but uses `workflow.status` which for these order is `completed` and when run calculates correctly for the examples I am pointing out.

So do we have two methods called something like `calculateReservedInventory` that uses different logic? I think you can probably see how that's problematic.

So is the implication here the `updateSimpleInventory` recalculates reserved inventory but `updateSimpleInventoryBulk` does not which is the issue I am pointing out. Are you implying that's by design? If so I can sort of see that but the copy and paste comment there seem to imply that it should.

The other issue here is that the concept of "Approving" a payment is used here. There are no payment systems that I know of that have an "Approval" step, that's completely an artificial Reaction conceit. So no system will ever have an analog where this event would make sense. I think this is done under the assumption that people will be using the Operator Tool to manually "approve" orders which I do not think is going to be the case for any Enterprise users. This should really be rethought. Almost every user of the system has requested that we either remove this step or have an "auto-approve" option.

That being said I *think* the short-term solution is for us to emit this phony `orderApproved` immediately on order placement, which is honestly a better solution for now since it's doesn't involve a change to core code.

**Comment by zenweasel** [comment]
I also notice that the Operator Tools have to way to Approve an order so this seems like it could never work no matter how you do it.

**Comment by aldeed** [comment]
@zenweasel We did hope to eventually make the "remove from stock" point customizable, but that so far isn't done. "Upon approving the last payment" was deemed to be the best point to do it until it's configurable. See https://docs.reactioncommerce.com/docs/core-plugins-simple-inventory#when-to-update-inventory

As for the various code paths, this was part of making it perform well. Both `updateSimpleInventoryBulk` and `updateSimpleInventory` call `recalculateReservedInventory`, which looks up every order that contains that item and isn't yet processed, adding up all of that inventory as the new "reserved" number. This is a brute force method that is necessary when the in stock qty changes or when you want to force recalc if the reserved number becomes wrong.

But during normal flows, we can avoid the expensive Orders lookup, so we do. This is where the event listening comes in. When an order payment is approved or order group is canceled or order is created, we increment or decrement the reserved or in-stock numbers as necessary based on the items in that order. The assumption is that this won't be foolproof, but it's also not mission-critical info, and there is a button in the UI to force recalculate if it gets really off.

@zenweasel Regarding the `Orders.find`, it does not include an approved payments check, but that is done within the `reduce` when it calls `orderIsApproved(order)`.

So long story short, I think if you are properly syncing `payment.status` to be "approved" (or anything other than "created"), then the next call to `updateSimpleInventoryBulk` should set the correct "reserved" number.

**Comment by aldeed** [comment]
Also, @kieckhafer worked on it last I think, but there should definitely be a way to approve order payments from the UI.

**Comment by zenweasel** [comment]
@aldeed Thanks also for the info, sorry this is taking me so long to unravel.

So is the intent here to only run the recalc when records are newly inserted? In my testing at least that is what appears to be happening? Is that on purpose? This statement makes it sounds like it should always run:
>  if you are properly syncing payment.status to be "approved" (or anything other than "created"), then the next call to updateSimpleInventoryBulk should set the correct "reserved" number.


Where is the button to recalc? Does that recalc all products or just an individual one?

**Comment by zenweasel** [comment]
@kieckhafer Can you confirm about approvals? Not that it exactly matters for us, but I only see "Capture Payment" nothing about Approval.

**Comment by aldeed** [comment]
@zenweasel 

Looking more closely. According to the comment, which seems to match the code, it is intending to call it only if a new SimpleInventory document was inserted. Same thing in the singular version of the function.

This seemed weird at first, but thinking it through, it makes sense. Why would updating the in stock quantity have any impact on the reserved qty? The same number are reserved even if you get 20 more in stock. Only placing, canceling, or fulfilling an order would change the reserved quantity.

There is a button titled "Recalculate reserved quantity" on the "Inventory" card for a sellable variant. You can only recalc for a single variant at a time in the UI. You could write a script to call the mutation repeatedly to sort of do it in bulk I guess.

**Comment by kieckhafer** [comment]
Approvals are "automatic" now. 

We were looking at removing the approval step completely, but at the time of refactoring all the orders work, we were also refactoring inventory and a few other things, and the `Approval` step was (is?) being used to perform actions outside of `Orders`, so I left it in, but also made it automatic with the `Capture`: https://github.com/reactioncommerce/reaction/blob/master/imports/plugins/core/orders/client/components/OrderPayments.js#L54:L54

Now that I'm looking at it, I don't know why this was done on the client instead of the server, refactoring it to be done on the server - or at this point completely removing the `approval` status - is probably a good idea.

Removing it may be easy at this point now that a lot of the other refactoring has been completed.

**Comment by aldeed** [comment]
@zenweasel After all discussion here, I think the main issue is, as @chrispotter speculated, that payment status is likely never getting sync'd from your external system, or is being done in such a way that the event isn't emitted. If payment status is moved beyond "created" and `afterOrderApprovePayment` is emitted, then reserved qty should update correctly.

Unfortunately, even doing that correctly going forward won't fix values that are currently incorrect. You'll likely need to call `recalculateReservedInventory` mutation for every variant/option once to reset, and then the normal process would keep things correct going forward. If you script something to periodically call that for every product as a failsafe, that will also help.

**Comment by zenweasel** [comment]
@kieckhafer @aldeed @chrispotter Thanks all for your help here

Just want to point out that pretty much every customer you encounter is going to have the same problem that SD Labs is having here. As I pointed out in another ticket, there currently is no way to trigger the "approval" step from server code, nor is it possible to fire the event directly in the payment code because the order has not been created yet. For almost all companies "Approval" won't be a thing and I think for a lot of companies "Capture" will also not be a thing.

We are going to work around it for now by just doing that same work manually (aka copy and pasting that same code into custom code) but I know you can all understand how that's going to be brittle..

I think another look at creating more detailed docs would really be a great boon to anybody trying to understand how this system works. It's a sophisticated system for a mission-critical task. I know this module got built in a great hurry largely for our benefit so if I/we can help let me/us know.

**Comment by aldeed** [comment]
@zenweasel The `approveOrderPayments` mutation could be called in server code. It does the approval and emits the event. It looks like it may not respect `isInternalCall`, though. 

Otherwise, points taken.

**Comment by zenweasel** [comment]
I know it can't because I tried. The `hasPermission` function is not available in a server context

**Comment by kieckhafer** [comment]
@zenweasel I'm not sure if I'm understanding your issue with `hasPermission` entirely, so this comment could be way off of what you're describing, but we do call the function `userHasPermission` on the context, not just `hasPermission`, perhaps you can try to call `context.userHasPermission` and see if that works?

https://github.com/reactioncommerce/reaction/blob/master/imports/node-app/core/util/buildContext.js#L36

**Comment by zenweasel** [comment]
@kieckhafer and worked together and confirmed that this function is NOT callable on the server side because `userHasPermission` is not available in the context from an event listener registered on startup.
